core/src/main/java/io/substrait/expression/proto/ExpressionProtoConverter.java

```diff
@@ -440,8 +440,8 @@ public Expression visit(io.substrait.expression.Expression.WindowFunctionInvocat
                         .build())
             .collect(java.util.stream.Collectors.toList());
 
-    Expression.WindowFunction.Bound upperBound = toBound(expr.upperBound());
-    Expression.WindowFunction.Bound lowerBound = toBound(expr.lowerBound());
+    Expression.WindowFunction.Bound upperBound = expr.upperBound().accept(TO_BOUND_VISITOR);
+    Expression.WindowFunction.Bound lowerBound = expr.lowerBound().accept(TO_BOUND_VISITOR);
```

Comment:
So you're putting an identical bounds visitor into both lowerBound and upperBound but call getLowerBound or getUpperBound as appropriate?  Why not have only one variable for storing bounds?

Reply:
I've cleaned this up a bit to hide the visit invocation behind a static method:
```java
    Expression.WindowFunction.Bound lowerBound = BoundConverter.convert(expr.lowerBound());
    Expression.WindowFunction.Bound upperBound = BoundConverter.convert(expr.upperBound());
```

To your other point, are you suggesting have something like a single `class WindowBounds` which contains both the lower and upper bound? The main reason for them to separate is to better match the [protos](https://github.com/substrait-io/substrait/blob/51765cccd92fa0014e9cf077cc8cc952731f33d0/proto/substrait/algebra.proto#L261-L263).

Another thing as well, is that the POJO <-> Proto conversion used to care about direction which is why there was a lower bound and upper bound variant. However, the substrait spec don't care about this, it was a Calcite detail. By pulling the direction out of the POJOs, I've limited the code that cares about direction to the POJO -> Calcite conversion in ExpressionRexConverter.

Reply:
Thanks for the changes, it looks less weird now.  Also agreed that the protos are oddly defined (by having multiple semantics in one field).